### PR TITLE
Migrate metrics agent layer to Struct API and bullet Stream

### PR DESCRIPTION
**Pull request by runesoerensen**
Migrate metrics agent layer to Struct API and bullet Stream. Taking over from #320 with a smaller and more modular process.

### CODE REVIEW

**Review by schneems — COMMENTED** [review]
I looked at all the commits. They're easy to follow in isolation. You were right. It was quite easy to review.

I left one comment about why I left some code in that seemingly did nothing, it was intentional. Basically I'm being defensive about guarding against possible breaking future modifications. I'm curious if that description makes sense and your thoughts on the relative costs/benefits.

**Comment by schneems** [comment]
I think it makes sense to keep iterating. I did some work on top for this layer then I'm pretty happy with it https://github.com/heroku/buildpacks-ruby/pull/322. And then I want to move over to porting other layers, basically re-doing the work of #320 in this style if that's agreeable.

**Comment by runesoerensen** [comment]
To be honest, I should probably not have included https://github.com/heroku/buildpacks-ruby/pull/321/commits/e0733ed20175c74c6362bbece0d0e3dde4d64999 on this branch. I wrote each commit as examples of focused *types of changes* that could be reviewed/integrated/merged better and faster independently (as separate PRs), and ended up confusing that by opening a PR including all the changes. To walk through my thought process (and noting things I could've done better in hindsight/based on your review):

1. https://github.com/heroku/buildpacks-ruby/pull/321/commits/6af2837b465c5d0fd85eb65c28ee9bd96c820409 would be the first commit (or PR, particularly if the same migration was applied to all the layers in a series of commits like https://github.com/heroku/buildpacks-jvm/pull/721/commits), along with https://github.com/heroku/buildpacks-ruby/pull/321/commits/e344c82d9265647d53ffe95b6734d613e494621c (bumping `libcnb.rs` to `0.23.0`).

    If I had opened a PR just bumping the libcnb version and migrating the metrics agent layer to the struct layer API, our lint checks would've prevented merging it and, among other things, forced me to make the changes required to allow using deprecated features (i.e. saving you from most of [this work](https://github.com/heroku/buildpacks-ruby/pull/323/files)).
2. The following three commits https://github.com/heroku/buildpacks-ruby/compare/e0733ed20175c74c6362bbece0d0e3dde4d64999~3...e0733ed20175c74c6362bbece0d0e3dde4d64999 were not part of the code changes in https://github.com/heroku/buildpacks-ruby/pull/320, or important to the end goal: Making the Ruby CNB's logging style consistent with our shared approach (in this case by adopting `bullet_stream`, facilitated by/necessitating the struct layer API migration). I still included them because of this comment in the original PR description: 

    > To minimize the change, I tried to avoid refactoring code as much as possible

    Which I totally get. Those commits were examples of "unrelated" code clean-up/maintenance/bug fixes I felt compelled to make while migrating to the struct layer API. They were "unnecessary" in the sense that the task could've been achieved without them, but not making these changes would also feel like a compromise, wasting the opportunity that revisiting code often presents.

    The good news is that we don't have to compromise and avoid such quick changes; introducing them in separate commits/PRs makes it extremely easy to review and integrate, and can help focus the conversation on the substance.

    Case in point, if each of these commits had been reviewed individually:

    * https://github.com/heroku/buildpacks-ruby/pull/321/commits/167cbb8b61b1614688b1052f290123f964400aab was a straightforward bug fix/improvement that would take around 60 secs or less to review and approve.
    * https://github.com/heroku/buildpacks-ruby/commit/56004d793c2f9ddf734d8e51a4810685f9920133 introduced a cleaner `Metadata` struct, but could've been further improved after your review, to factor in the [SPOT considerations here](https://github.com/heroku/buildpacks-ruby/pull/322/commits/6fd59ee83712492f3a27195a2f1e5a31dde0d860).
    * https://github.com/heroku/buildpacks-ruby/commit/e0733ed20175c74c6362bbece0d0e3dde4d64999 would have lead to this highly relevant [comment (that lead me to write this response)](https://github.com/heroku/buildpacks-ruby/pull/321/commits/6af2837b465c5d0fd85eb65c28ee9bd96c820409#r1768974703), but specifically related to that part alone. That being said, it'd definitely have been better if I had introduced this particular change *after* migrating to `bullet_stream`.
    
        As you can see from my commit message, I was very much open to the possibility that I was missing something obvious - and the change would at best have zero functional impact. By making the change before the `bullet_stream` commit, I unnecessarily made it more difficult to discard in case it didn't pass code review (as rebasing/merging without that commit would've caused conflicts).
3. Finally, [my last commit (migrating to `bullet_stream`)](https://github.com/heroku/buildpacks-ruby/pull/321/commits/b4b1fcfbc178845474741b02005c8c277899f21e) would actually have benefitted from the first PR/commit described above (in particular, passing all the lints). When I committed that change, I had hundreds of clippy deprecation warnings - so I disregarded them, causing me to the miss the ones that actually mattered here (which you fixed [in this commit](https://github.com/heroku/buildpacks-ruby/pull/323/commits/744a3c5194ffb50952b2eee856b395164bd5358f)).

I've talked a lot about PRs here, but hope it's also clear what I really care about: Clean, focused, atomic commits, that are able to stand on its own (e.g. be reviewed and tested independently, without introducing regressions) and make just one type of change (and never more than one logical change). GitHub PRs can be a great tool in that effort, not least because they trigger CI workflows that can help during development. But as far as I'm concerned, if the commits make sense on their own, you can basically structure them however you like in PRs.

At any rate: Sounds like you want to iterate based on this branch -- I've approved the two PRs building on it, but haven't merged anything. Feel free to proceed however you prefer :)

File: buildpacks/ruby/src/main.rs [hunk 1]
```diff
@@ -134,16 +134,15 @@ impl Buildpack for RubyBuildpack {
...
+                    &context,
+                    section.as_ref(),
                 )?;
-
                 (
                     section.end_section(),
-                    layer_data.env.apply(Scope::Build, &env),
+                    // TODO: This should likely be removed (also eliminating the need for the `handle_metrics_agent_layer` function to return a `LayerRef`)?
+                    // Included the logic here for reference, but it doesn't seem necessary (the layer env isn't explicitly modified as far as I can tell, but perhaps there's some trait API magic happening I'm not familiar with?).
+                    layer_ref.read_env()?.apply(Scope::Build, &env),
```
[the remark this answers is not in the document]
Comment (runesoerensen) [reply buildpacks/ruby/src/main.rs]: > it's easier for me to standardize on an interface for all my layer calls so they all do the same thing, versus difficult to discover action-at-a-distance problems in the future.

I want a standard model/interface with less risk of *spooky* *"action-at-a-distance"* too! The best way to do that, in this case, would eliminate any actions affecting any variables altogether though. So https://github.com/heroku/buildpacks-ruby/pull/321/commits/e0733ed20175c74c6362bbece0d0e3dde4d64999 should probably also have moved https://github.com/heroku/buildpacks-ruby/blob/e0733ed20175c74c6362bbece0d0e3dde4d64999/buildpacks/ruby/src/main.rs#L122 below the metrics agent layer logic - this way, it'd be obvious that the `env` variable couldn't possibly be affected by any distant/preceding actions.

I might be biased, coming from Copenhagen and all, but I think even a certain German scientist may agree with this interpretation :)

**Pull request merged**